src/libraries/System.Private.CoreLib/gen/EventSourceGenerator.Parser.cs

```diff
@@ -16,58 +17,47 @@ namespace Generators
-            {
-                _compilation = compilation;
-                _cancellationToken = cancellationToken;
-                _reportDiagnostic = reportDiagnostic;
-            }
-
-            public EventSourceClass[] GetEventSourceClasses(List<ClassDeclarationSyntax> classDeclarations)
+            public static ImmutableArray<EventSourceClass> GetEventSourceClasses(ImmutableArray<ClassDeclarationSyntax> classDeclarations, Compilation compilation, CancellationToken cancellationToken)
             {
-                INamedTypeSymbol? autogenerateAttribute = _compilation.GetBestTypeByMetadataName("System.Diagnostics.Tracing.EventSourceAutoGenerateAttribute");
+                INamedTypeSymbol? autogenerateAttribute = compilation.GetBestTypeByMetadataName("System.Diagnostics.Tracing.EventSourceAutoGenerateAttribute");
                 if (autogenerateAttribute is null)
                 {
                     // No EventSourceAutoGenerateAttribute
-                    return Array.Empty<EventSourceClass>();
+                    return ImmutableArray<EventSourceClass>.Empty;
                 }
 
-                INamedTypeSymbol? eventSourceAttribute = _compilation.GetBestTypeByMetadataName("System.Diagnostics.Tracing.EventSourceAttribute");
+                INamedTypeSymbol? eventSourceAttribute = compilation.GetBestTypeByMetadataName("System.Diagnostics.Tracing.EventSourceAttribute");
                 if (eventSourceAttribute is null)
                 {
                     // No EventSourceAttribute
-                    return Array.Empty<EventSourceClass>();
+                    return ImmutableArray<EventSourceClass>.Empty;
                 }
 
-                List<EventSourceClass>? results = null;
+                ImmutableArray<EventSourceClass>.Builder? results = null;
                 // we enumerate by syntax tree, to minimize the need to instantiate semantic models (since they're expensive)
-                foreach (IGrouping<SyntaxTree, ClassDeclarationSyntax>? group in classDeclarations.GroupBy(x => x.SyntaxTree))
+                foreach (IGrouping<SyntaxTree, ClassDeclarationSyntax> group in classDeclarations.GroupBy(static x => x.SyntaxTree))
                 {
                     SemanticModel? sm = null;
                     EventSourceClass? eventSourceClass = null;
-                    foreach (ClassDeclarationSyntax? classDef in group)
+                    foreach (ClassDeclarationSyntax classDef in group)
                     {
-                        if (_cancellationToken.IsCancellationRequested)
+                        if (cancellationToken.IsCancellationRequested)
                         {
                             // be nice and stop if we're asked to
-                            return results?.ToArray() ?? Array.Empty<EventSourceClass>();
+                            return results?.ToImmutable() ?? ImmutableArray<EventSourceClass>.Empty;
```

Comment:
```suggestion
                            cancellationToken.ThrowIfCancellationRequested();
```

The generator driver will handle cancellation exceptions for you and keep the cached values. Returning empty results counts as a 'success' run, and the driver will cache the empty results, meaning you need to do more work next time around.